tasks/runtime.yml

```diff
@@ -20,27 +20,55 @@
     mode: 0755
     state: directory
 
-- name: Generate a Munge key for the platform
+# Munge key handling is slightly complex as trying to cover cases with the following in the play:
+# a. entire cluster (standard ansible deployment)
+# b. only control node (part of image-based deploy)
+# c. only single compute node (for image building)
+# Therefore cannot depend on being able to generate a key on the control node and slurp + redistribute.
+# A local key can be generated from the path {{ openhpc_munge_key }}, 
+
+- name: Write Munge key from local disk
+  copy:
+    src: "{{ openhpc_munge_key }}"
+    dest: "/etc/munge/munge.key"
+    owner: munge
+    group: munge
+    mode: 0400
+  when: openhpc_munge_key is defined
+  notify:
+    - Restart Munge service
+
+- name: Generate a Munge key
+  # NB this is usually a no-op as the package install actually generates a (node-unique) one, so won't usually trigger handler
   command: "dd if=/dev/urandom of=/etc/munge/munge.key bs=1 count=1024"
   args:
     creates: "/etc/munge/munge.key"
-  when: inventory_hostname == openhpc_slurm_control_host
   notify:
     - Restart Munge service
 
-- name: Retrieve Munge key from Slurm control host
+- name: Retrieve Munge key
   slurp:
     src: "/etc/munge/munge.key"
   register: slurm_munge_key
-  when: inventory_hostname == openhpc_slurm_control_host
 
-- name: Write Munge key
+- name: Save Munge key locally if single node play
+  fetch:
+    src: "/etc/munge/munge.key"
+    dest: '.'
```

Comment:
I assume this is so you can then reuse it for the next image build? Could it make sense to generate it on the local machine?

Reply:
Yes, again allows only-build-headnode, and only-build-compute-image to be the same. We could generate it locally, but I sort of like the fact that if you're in build-an-entire-cluster mode you don't get a key left lying around on the ansible controlhost.